pkg/mutation/system_external_data.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	responses, err := s.sendRequests(providerKeys)
+	if err != nil {
+		return err
+	}
+
+	if err := s.mutateWithExternalData(obj, responses); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// sendRequests sends requests to all providers in parallel.
+func (s *System) sendRequests(providerKeys map[string]sets.String) (map[string]map[string]*externaldata.Item, error) {
+	responses := make(map[string]map[string]*externaldata.Item)
+	grp, _ := errgroup.WithContext(context.Background())
+	for name, keys := range providerKeys {
+		provider, err := s.providerCache.Get(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get external data provider %s: %w", name, err)
+		}
+
+		fn := s.sendRequestToExternalDataProvider
+		if fn == nil {
+			fn = defaultSendRequestToExternalDataProvider
+		}
+
+		provider, keys := provider, keys
+		grp.Go(func() error {
+			resp, err := fn(context.Background(), &provider, keys.List())
+			if err != nil {
+				return fmt.Errorf("failed to send request to external data provider %s: %w", provider.Name, err)
+			}
+			if err := validateExternalDataResponse(resp); err != nil {
+				return fmt.Errorf("failed to validate response from external data provider %s: %w", provider.Name, err)
+			}
+
+			responses[provider.Name] = make(map[string]*externaldata.Item)
+			for _, item := range resp.Response.Items {
+				item := item // for scoping
+				responses[provider.Name][item.Key] = &item
+			}
+
+			return nil
+		})
+	}
+	if err := grp.Wait(); err != nil {
+		return nil, err
+	}
+
+	return responses, nil
+}
+
+// mutateWithExternalData recursively traverses the given object and replaces
+// all external data placeholders with the corresponding external data items.
+func (s *System) mutateWithExternalData(object *unstructured.Unstructured, externalData map[string]map[string]*externaldata.Item) error {
+	var mutate func(interface{}) []error
+	mutate = func(current interface{}) []error {
+		var allErrors []error
+		switch obj := current.(type) {
+		case map[string]interface{}:
+			for k, v := range obj {
+				placeholder, ok := types.ToExternalDataPlaceholder(v)
+				if !ok {
+					// not a placeholder, let's continue recursing
+					allErrors = append(allErrors, mutate(v)...)
+					continue
+				}
+
+				// our base case - we found a placeholder and we should resolve it
+				providerResponse, ok := externalData[placeholder.ProviderName]
+				if !ok {
+					allErrors = append(allErrors, fmt.Errorf("failed to find external data provider %s in responses", placeholder.ProviderName))
+					continue
+				}
+
+				mutator := s.Get(*placeholder.MutatorID)
+				if mutator == nil {
+					allErrors = append(allErrors, fmt.Errorf("failed to find mutator %s", placeholder.MutatorID))
+					continue
+				}
+
+				// mutator helps us resolve the external data response because
+				// it holds the failure policy and the default value
+				value, err := mutator.ResolveExternalData(providerResponse[placeholder.ValueAtLocation])
```

Comment:
Does this need to call the corresponding mutator? Can we mutate the object directly in this function?

If so, we simplify the mutator interface and avoid some code duplication.

We'd probably need to cache the external data object in `Placeholder{}` in order to do things like resolve failure conditions.